src/posts/xarray-kvikio/index.md

```diff
@@ -0,0 +1,179 @@
+
+For Xarray, the key bit is that kvikIO exposes a zarr store [`kvikio.zarr.GDSStore`](https://docs.rapids.ai/api/kvikio/stable/api.html#zarr) that does all the hard work for us. Since Xarray knows how to read Zarr stores, we can adapt it to create a new storage backend that uses `kvikio`. And thanks to recent work funded by the [Chan Zuckerberg Initiative](https://xarray.dev/blog/czi-eoss-grant-conclusion), creating and registering a [new backend](https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html) is quite easy!
+
+## Integrating with Xarray
+
+Getting all this to work nicely requires using three in-progress pull requests that
+
+1. [Teach Zarr to handle alternative array classes](https://github.com/zarr-developers/zarr-python/pull/934)
+2. [Rewrite a small bit of Xarray to not cast all data to a numpy array after read from disk](https://github.com/pydata/xarray/pull/6874)
+3. [Make a backend that connects Xarray to kvikIO](https://github.com/xarray-contrib/cupy-xarray/pull/10)
+
+Writing the backend for Xarray was relatively easy with most of the code copied over or inherited from the existing Zarr backend. We did have to ensure that dimension coordinates could be read in directly to host memory without raising an error (by default kvikIO loads all data to device memory). This is required because Xarrays creates `pandas.Index` objects for such variables. In the future, we could consider using `cudf.Index` instead to allow a fully GPU-backed Xarray object.
+
+## Usage
+
+Assuming you have all the pieces together (see [Appendix I](#appendix-i--step-by-step-install-instructions) and [Appendix II](#appendix-ii--making-sure-gds-is-working) for step-by-step instructions), then using all this cool technology only requires adding `engine="kvikio"` to your `open_dataset` line (!)
+
+```python
+import xarray as xr
+
+ds = xr.open_dataset("file.zarr", engine="kvikio", consolidated=False)
+```
+
+Notice that importing `cupy_xarray` was not needed. `cupy_xarray` uses [entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/) to register the kvikIO backend with Xarray.
+
+With this `ds.load()` will load directly to GPU memory and `ds` will now contain CuPy arrays. At present there are a few limitations:
+
+1. Zarr stores cannot be read with consolidated metadata, and
+2. compression is unsupported by the kvikIO backend.
+
+## Quick demo
+
+First create an example uncompressed dataset to read from
+
+```
+import xarray as xr
+
+store = "./air-temperature.zarr"
+
+airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")
+
+for var in airt.variables:
+    airt[var].encoding["compressor"] = None
+airt.to_zarr(store, mode="w", consolidated=True)
+```
+
+Now read
+
+```
+# consolidated must be False
+ds = xr.open_dataset(store, engine="kvikio", consolidated=False)
+ds.air
+```
+
+```
+<xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)>
```

Comment:
Not sure if syntax highlighting helps here, but won't hurt I guess.
```suggestion
```python
<xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)>
```